### PR TITLE
Runtime support for static virtual interface methods

### PR DESCRIPTION
**Pull request by trylek**
This change implements initial CoreCLR runtime support for static virtual interface methods and adds over 1700 test cases covering various aspects of this new runtime feature.

1) In the JIT, we are relaxing the assumption that ".constrained" must be always followed by a "callvirt" by allowing two more IL instructions to be constrained, "call" and "ldftn".

2) In the JIT interface, we're adding bits of code to CEEInfo::getCallInfo to properly handle the new static virtual methods. We're extending constrained method resolution to cater for the static virtual method case in a new method "ResolveStaticVirtualMethod".

3) In our work on the implementation we found a pre-existing JIT interface issue - the interplay between getCallInfo and embedGenericHandle doesn't work well in case of constrained calls as the existing API surface doesn't provide any means for communicating compile-time resolution of the constraint from getCallInfo to embedGenericHandle. In this change we're working around this by deferring to runtime resolution of the constraint in the presence of shared generics.

4) In the method table builder, we're newly tracking a flag saying whether we're implementing an interface declaring static virtual methods (fHasVirtualStaticMethods) that is used to speed up the runtime checks by skipping irrelevant interfaces.

5) For Crossgen / Crossgen2, this change only adds minimalistic support in the form of bailing out in getCallInfo whenever we hit a call to a static virtual method, cancelling AOT compilation of the caller and deferring it to runtime JIT.

### CODE REVIEW

**Comment by davidwrighton** [comment]
@jkotas  This is a complete implementation of the static virtuals feature, I'd like you opinion as to whether or not we should wrap all the changes in a #ifdef FEATURE_VIRTUAL_STATIC_INTERFACE_METHODS wrapper or not.

**Comment by jkotas** [comment]
> we should wrap all the changes in a #ifdef FEATURE_VIRTUAL_STATIC_INTERFACE_METHODS wrapper or not.

I do not think we need to wrap these changes under ifdefs. Ifdefs for FEATUREs that are enabled in all shipping configs have no value. 

Also, we may want to delete similar existing `FEATURE_DEFAULT_INTERFACES` when we get a chance.

**Comment by davidwrighton** [comment]
@jkotas that matches my opinion, but it does differ from past approaches, so I wanted your opinion.

**Comment by trylek** [comment]
The library test bug seems to be https://github.com/dotnet/runtime/issues/52710.

File: src/tests/Common/dirs.proj [hunk 1]
```diff
@@ -14,6 +14,8 @@
       <DisabledProjects Include="$(TestRoot)GC\Performance\Framework\GCPerfTestFramework.csproj" />
       <DisabledProjects Include="$(TestRoot)Loader\classloader\generics\regressions\DD117522\Test.csproj" />
       <DisabledProjects Include="$(TestRoot)Loader\classloader\generics\GenericMethods\VSW491668.csproj" /> <!-- issue 5501 -->
+      <DisabledProjects Include="$(TestRoot)Loader\classloader\StaticVirtualMethods\**\generatetest.csproj" /> <!-- test generators -->
+      <DisabledProjects Include="$(TestRoot)Performance\Scenario\JitBench\unofficial_dotnet\JitBench.csproj" /> <!-- no official build support for SDK-style netcoreapp2.0 projects -->
```
Comment (trylek) [inline src/tests/Common/dirs.proj:18]: I don't know why JitBench.csproj seems like our addition, I guess I messed up the conflict merge.

**Review by davidwrighton — APPROVED** [review]
Looks good to me.

**Pull request merged**